drivers/clock_control/clock_control_r8a7795_cpg_mssr.c

```diff
@@ -20,140 +20,285 @@
 
 LOG_MODULE_DECLARE(clock_control_rcar);
 
+#define R8A7795_CLK_SD_STOP_BIT BIT(8)
+#define R8A7795_CLK_SD_DIV_MASK 0x3
+#define R8A7795_CLK_SD_DIV_SHIFT 0
+
+#define R8A7795_CLK_SDH_STOP_BIT BIT(9)
+#define R8A7795_CLK_SDH_DIV_MASK 0x7
+#define R8A7795_CLK_SDH_DIV_SHIFT 2
+
+#define R8A7795_CLK_CANFD_STOP_BIT BIT(8)
+#define R8A7795_CLK_CANFD_DIV_MASK 0x3f
 
 struct r8a7795_cpg_mssr_config {
 	mm_reg_t base_address;
 };
 
-int r8a7795_cpg_core_clock_endisable(uint32_t base_address, uint32_t module,
-				     uint32_t rate, bool enable)
+struct r8a7795_cpg_mssr_data {
+	struct rcar_cpg_mssr_data cmn;
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table core_props[] = {
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_S3D4, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_KHZ(66600)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD0H, 0x0074, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD0, 0x0074, R8A7795_CLK_SD0H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD1H, 0x0078, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD1, 0x0078, R8A7795_CLK_SD1H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD2H, 0x0268, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD2, 0x0268, R8A7795_CLK_SD2H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD3H, 0x026C, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD3, 0x026C, R8A7795_CLK_SD3H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_CANFD, 0x0244, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_S0D12, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_KHZ(66600)),
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table mod_props[] = {
+	RCAR_MOD_CLK_INFO_ITEM(310, R8A7795_CLK_S3D4),
+
+	RCAR_MOD_CLK_INFO_ITEM(311, R8A7795_CLK_SD3),
+	RCAR_MOD_CLK_INFO_ITEM(312, R8A7795_CLK_SD2),
+	RCAR_MOD_CLK_INFO_ITEM(313, R8A7795_CLK_SD1),
+	RCAR_MOD_CLK_INFO_ITEM(314, R8A7795_CLK_SD0),
+};
+
+static int r8a7795_cpg_enable_disable_core(const struct r8a7795_cpg_mssr_config *cfg,
+					struct cpg_clk_info_table *clk_info,
+					uint32_t enable)
 {
-	uint32_t divider;
-	unsigned int key;
 	int ret = 0;
+	uint32_t reg;
 
-	/* Only support CANFD core clock at the moment */
-	if (module != R8A7795_CLK_CANFD) {
-		return -EINVAL;
+	enable = !!enable;
+
+	switch (clk_info->module) {
+	case R8A7795_CLK_SD0:
+	case R8A7795_CLK_SD1:
+	case R8A7795_CLK_SD2:
+	case R8A7795_CLK_SD3:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A7795_CLK_SD_STOP_BIT;
+		reg |= !enable * R8A7795_CLK_SD_STOP_BIT;
```

Comment:
This is... quite unconventional. Are you sure you didn't meant `(!enable << R8A7795_CLK_SD_STOP_BIT)` ?

Reply:
Okay, I see what you did there wish you luck with external review.